File: 0015 - 3Sum/solution.cpp
class Solution {
public:
    vector<vector<int>> threeSum(vector<int>& nums) {
        int n = nums.size();
        vector<vector<int>> res;
        sort(nums.begin(), nums.end());
        for(int i = 0; i < n; i++){
            if(i>0 && nums[i]==nums[i-1]) continue;
            int l = i+1, r = n-1;
            while(l < r){
                int L = nums[l], R = nums[r], s = nums[i]+L+R;
                if(s == 0){
                    res.push_back({nums[i], L, R});
                    while(l<r && nums[l]==L) l++;
                    while(l<r && nums[r]==R) r--;
                }
                else s<0?(l++):(r--);
            }
        }
        return res;
    }
};

File: 0015 - 3Sum/solution.py
class Solution:
    def threeSum(self, nums: List[int]) -> List[List[int]]:
        from itertools import combinations
        res, n, p, z = set(), [], [], 0
        for i in nums:
            if i < 0:
                n.append(i)
            elif i > 0:
                p.append(i)
            else:
                z += 1
        N, P = set(n), set(p)
        if z:
            for i in n:
                if -i in P:
                    res.add((-i, 0, i))
            if z >= 3:
                res.add((0, 0, 0))
        for x, y in combinations(n, 2):
            if -x-y in P:
                res.add(tuple(sorted([x, y, -x-y])))
        for x, y in combinations(p, 2):
            if -x-y in N:
                res.add(tuple(sorted([x, y, -x-y])))
        return map(list, res)